### PR TITLE
Laboratory work #7, Zakirova Karina - 20FPL1 

### PR DESCRIPTION
**Pull request by KarinaZakirova**

### CODE REVIEW

File: lab_7_llm/settings.json [hunk 1]
```diff
@@ -1,8 +1,8 @@
 {
   "parameters": {
-    "model": "",
-    "dataset": "",
-    "metrics": []
+    "model": "cointegrated/rubert-base-cased-nli-threeway",
+    "dataset": "cointegrated/nli-rus-translated-v2021",
+    "metrics": ["accuracy"]
   },
   "target_score": 0
```
Comment (SalnikovIgor) [inline lab_7_llm/settings.json]: What grade do you expect?

File: lab_7_llm/main.py [hunk 2]
```diff
@@ -41,6 +45,8 @@ def obtain(self) -> None:
         Raises:
             TypeError: In case of downloaded dataset is not pd.DataFrame
         """
+        self._raw_data = load_dataset('cointegrated/nli-rus-translated-v2021', split='train'). \
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: Use `dev` split

File: lab_7_llm/main.py [hunk 3]
```diff
@@ -55,12 +61,31 @@ def analyze(self) -> dict:
...
+                'dataset_sample_min_len': len(min(self._raw_data["premise_ru"])),
+                'dataset_sample_max_len': len(max(self._raw_data["premise_ru"]))
+                }
 
     @report_time
     def transform(self) -> None:
         """
         Apply preprocessing transformations to the raw dataset.
         """
+        # TODO: ask about «Map target with class labels»
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: Delete

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]

**Comment by SalnikovIgor** [comment]
Fix remaining comments and requirements test

File: lab_7_llm/main.py (hunk 3 above)
Comment (KarinaZakirova) [reply lab_7_llm/main.py]: done

File: lab_7_llm/settings.json (hunk 1 above)
Comment (KarinaZakirova) [reply lab_7_llm/settings.json]: done

File: lab_7_llm/main.py [hunk 4]
```diff
@@ -136,6 +168,20 @@ def analyze_model(self) -> dict:
         Returns:
             dict: Properties of a model
         """
+        tensor_data = torch.ones(1, 512, dtype=torch.long)
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: 512 - "magic number". Take this parameter from `model.config`

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]
Fix one comment and implement the sample inference

File: lab_7_llm/main.py (hunk 4 above)
Comment (KarinaZakirova) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py [hunk 5]
```diff
@@ -41,6 +48,9 @@ def obtain(self) -> None:
         Raises:
             TypeError: In case of downloaded dataset is not pd.DataFrame
         """
+        self._raw_data = (load_dataset('cointegrated/nli-rus-translated-v2021', split='dev')
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: Use a variable  `self._hf_name` to pass the dataset name

File: lab_7_llm/main.py [hunk 6]
```diff
@@ -94,6 +128,7 @@ def __getitem__(self, index: int) -> tuple[str, ...]:
         Returns:
             tuple[str, ...]: The item to be received
         """
+        return "|".join((self._data.iloc[index]["premise"], self._data.iloc[index]["hypothesis"])),
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: Why join? You need to take a data using a `iloc`, then take the required columns, convert the values ​​to strings and return a tuple of them

File: lab_7_llm/main.py [hunk 7]
```diff
@@ -148,6 +194,12 @@ def infer_sample(self, sample: tuple[str, ...]) -> str | None:
         Returns:
             str | None: A prediction
         """
+        tokenizer = AutoTokenizer.from_pretrained(self._model_name)
+        tokens = tokenizer(sample[0].split("|")[0], sample[0].split("|")[1], return_tensors='pt')
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: Replace with tuple

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]

File: lab_7_llm/main.py (hunk 5 above)
Comment (KarinaZakirova) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py (hunk 6 above)
Comment (KarinaZakirova) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py (hunk 7 above)
Comment (KarinaZakirova) [reply lab_7_llm/main.py]: done

File: requirements.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+--extra-index-url https://download.pytorch.org/whl/cpu
+datasets==2.16.1
+evaluate==0.4.1
+pandas==2.1.4
+pydantic==2.5.3
```
Comment (SalnikovIgor) [inline requirements.txt:6]: Where do we use this dependency?

File: requirements.txt [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+pytest==7.4.4
+scikit-learn==1.4.0
+torch==2.1.2
+torchinfo==1.8.0
+tqdm==4.66.1
```
Comment (SalnikovIgor) [inline requirements.txt]: And this one?

File: lab_7_llm/main.py [hunk 10]
```diff
@@ -193,3 +274,11 @@ def run(self) -> dict | None:
...
             dict | None: A dictionary containing information about the calculated metric
         """
+        data = pd.read_csv(self._data_path).drop(columns=['Unnamed: 0'])
+
+        for metric in self._metrics:
+            if metric.value == "accuracy":
+                accuracy_metric = load(metric.value).compute(references=data['target'],
+                                                             predictions=data['predictions'])
+                return dict(accuracy_metric)
+        return None
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: Delete.  Python function returns by default `None`

File: lab_7_llm/main.py [hunk 11]
```diff
@@ -193,3 +274,11 @@ def run(self) -> dict | None:
         Returns:
             dict | None: A dictionary containing information about the calculated metric
         """
+        data = pd.read_csv(self._data_path).drop(columns=['Unnamed: 0'])
+
+        for metric in self._metrics:
+            if metric.value == "accuracy":
+                accuracy_metric = load(metric.value).compute(references=data['target'],
+                                                             predictions=data['predictions'])
+                return dict(accuracy_metric)
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py]: Isn't the `accuracy_metric` a dictionary anyway?

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]
Fix minor comments and when do you plan to implement a service for a score of 10?

File: lab_7_llm/main.py (hunk 11 above)
Comment (KarinaZakirova) [reply lab_7_llm/main.py]: I had problems with Mypy checks, so I've written it such way but now I've changed this function and delete dict() method

done

File: lab_7_llm/main.py (hunk 10 above)
Comment (KarinaZakirova) [reply lab_7_llm/main.py]: Mypy checks made me do this, but now I've changed this function

done

File: requirements.txt (hunk 9 above)
Comment (KarinaZakirova) [reply requirements.txt]: sorry, it is really extra 
done

File: requirements.txt (hunk 8 above)
Comment (KarinaZakirova) [reply requirements.txt:6]: it is necessary for implementation of the class Query in service.py

File: lab_7_llm/assets/main.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        let question = premise.value.concat("|", hypothesis.value)
+        await fetch("/infer", {
+            method: "POST",
+            headers: {
+                "Content-type": "application/json"
+            },
+            body: JSON.stringify({
+                question: question
+            })
+        }).then(response => {
```
Comment (SalnikovIgor) [inline lab_7_llm/assets/main.js]: `async/await` + `then`: very interesting what will happen in the event loop

File: lab_7_llm/main.py [hunk 13]
```diff
@@ -32,6 +28,7 @@ class RawDataImporter(AbstractRawDataImporter):
     """
     A class that imports the HuggingFace dataset.
     """
+    _raw_data: DataFrame
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py:31]: This parameter is already created in the class `AbstractRawDataImporter`

File: lab_7_llm/main.py [hunk 14]
```diff
@@ -41,6 +38,13 @@ def obtain(self) -> None:
         Raises:
             TypeError: In case of downloaded dataset is not pd.DataFrame
         """
+        self._raw_data = (load_dataset(self._hf_name, split='dev')
+                          .filter(lambda dataset: dataset['source'] == 'mnli')
+                          .to_pandas())
+
+    @property
+    def raw_data(self) -> DataFrame:
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py:46]: Also this

File: requirements.txt (hunk 8 above)
Comment (SalnikovIgor) [reply requirements.txt:6]: Ok😑

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]

**Pull request closed without merging**